### ISSUE TITLE
[ yesod ] Build fail's with ghc-9.0.1

### ISSUE THREAD

**Issue by mimi1vx** [yesodweb/yesod#1718]
```
[   87s] + ./Setup configure --prefix=/usr --libdir=/usr/lib64 --docdir=/usr/share/licenses/ghc-yesod '--libsubdir=$compiler/$pkgid' '--datasubdir=$pkgid' --ghc '--dynlibdir=/usr/lib64/$compiler/$pkgid' --global '--ghc-options= -optc-O2 -optc-Wall -optc-D_FORTIFY_SOURCE=2 -optc-fstack-protector-strong -optc-funwind-tables -optc-fasynchronous-unwind-tables -optc-fstack-clash-protection -optc-Werror=return-type ' -v -p --enable-shared --htmldir=/usr/share/doc/packages/ghc/html/libraries/yesod-1.6.1.0 --global
[   88s] Using Parsec parser
[   88s] Configuring yesod-1.6.1.0...
[   88s] Dependency aeson: using aeson-1.5.6.0
[   88s] Dependency base >=4.10 && <5: using base-4.15.0.0
[   88s] Dependency bytestring: using bytestring-0.10.12.1
[   88s] Dependency conduit >=1.3: using conduit-1.3.4.1
[   88s] Dependency data-default-class: using data-default-class-0.1.2.0
[   88s] Dependency directory: using directory-1.3.6.1
[   88s] Dependency fast-logger: using fast-logger-3.0.3
[   88s] Dependency file-embed: using file-embed-0.0.13.0
[   88s] Dependency monad-logger: using monad-logger-0.3.36
[   88s] Dependency shakespeare: using shakespeare-2.0.25
[   88s] Dependency streaming-commons: using streaming-commons-0.2.2.1
[   88s] Dependency template-haskell: using template-haskell-2.17.0.0
[   88s] Dependency text: using text-1.2.4.1
[   88s] Dependency unix: using unix-2.7.2.2
[   88s] Dependency unordered-containers: using unordered-containers-0.2.13.0
[   88s] Dependency wai >=1.3: using wai-3.2.3
[   88s] Dependency wai-extra >=1.3: using wai-extra-3.1.6
[   88s] Dependency wai-logger: using wai-logger-2.3.6
[   88s] Dependency warp >=1.3: using warp-3.3.14
[   88s] Dependency yaml >=0.8.17: using yaml-0.11.5.0
[   88s] Dependency yesod-core >=1.6 && <1.7: using yesod-core-1.6.18.8
[   88s] Dependency yesod-form >=1.6 && <1.7: using yesod-form-1.6.7
[   88s] Dependency yesod-persistent >=1.6 && <1.7: using yesod-persistent-1.6.0.5
[   88s] Source component graph: component lib
[   88s] Configured component graph:
[   88s]     component yesod-1.6.1.0-L1ZrEfHHjEL4ZPg5gUFJ40
[   88s]         include aeson-1.5.6.0-DKRYsoqDhrXJNujTZrDHEf
[   88s]         include base-4.15.0.0
[   88s]         include bytestring-0.10.12.1
[   88s]         include conduit-1.3.4.1-5jKrafZ5jYhKoJdTRMFD84
[   88s]         include data-default-class-0.1.2.0-FEqadazbCfKLPgsOi8Mq6U
[   88s]         include directory-1.3.6.1
[   88s]         include fast-logger-3.0.3-FaKeK891u6TFBhkVRPQKLr
[   88s]         include file-embed-0.0.13.0-4AN2M3x2UWSEsGAFWu7ZEd
[   88s]         include monad-logger-0.3.36-FL2BtKw8EeaF4wZmehPP8a
[   88s]         include shakespeare-2.0.25-4IGNqPAl1f13jGkgGXg0N5
[   88s]         include streaming-commons-0.2.2.1-Hw3qCB3irjv4km2jUH6Jg0
[   88s]         include template-haskell-2.17.0.0
[   88s]         include text-1.2.4.1
[   88s]         include unix-2.7.2.2
[   88s]         include unordered-containers-0.2.13.0-LSqNoVnVBinKYmInNBh54X
[   88s]         include wai-3.2.3-HxoYdmypAXX1MnCeAE2iqq
[   88s]         include wai-extra-3.1.6-LudkXIwcaQS8Nqvm3IxzCV
[   88s]         include wai-logger-2.3.6-I6z9W9PdRA25Ruoe0vixqu
[   88s]         include warp-3.3.14-2I6laE4MgmWJYarfgyelkn
[   88s]         include yaml-0.11.5.0-G8RzVbeSOqB50sCYmn2TZ
[   88s]         include yesod-core-1.6.18.8-6NbLNd2S7d5Bf3HJp6rKid
[   88s]         include yesod-form-1.6.7-LC9WO26D5bjC0MNFROBEQe
[   88s]         include yesod-persistent-1.6.0.5-9cO5dk1kmHOAsa4tLTj0r1
[   88s] Linked component graph:
[   88s]     unit yesod-1.6.1.0-L1ZrEfHHjEL4ZPg5gUFJ40
[   88s]         include aeson-1.5.6.0-DKRYsoqDhrXJNujTZrDHEf
[   88s]         include base-4.15.0.0
[   88s]         include bytestring-0.10.12.1
[   88s]         include conduit-1.3.4.1-5jKrafZ5jYhKoJdTRMFD84
[   88s]         include data-default-class-0.1.2.0-FEqadazbCfKLPgsOi8Mq6U
[   88s]         include directory-1.3.6.1
[   88s]         include fast-logger-3.0.3-FaKeK891u6TFBhkVRPQKLr
[   88s]         include file-embed-0.0.13.0-4AN2M3x2UWSEsGAFWu7ZEd
[   88s]         include monad-logger-0.3.36-FL2BtKw8EeaF4wZmehPP8a
[   88s]         include shakespeare-2.0.25-4IGNqPAl1f13jGkgGXg0N5
[   88s]         include streaming-commons-0.2.2.1-Hw3qCB3irjv4km2jUH6Jg0
[   88s]         include template-haskell-2.17.0.0
[   88s]         include text-1.2.4.1
[   88s]         include unix-2.7.2.2
[   88s]         include unordered-containers-0.2.13.0-LSqNoVnVBinKYmInNBh54X
[   88s]         include wai-3.2.3-HxoYdmypAXX1MnCeAE2iqq
[   88s]         include wai-extra-3.1.6-LudkXIwcaQS8Nqvm3IxzCV
[   88s]         include wai-logger-2.3.6-I6z9W9PdRA25Ruoe0vixqu
[   88s]         include warp-3.3.14-2I6laE4MgmWJYarfgyelkn
[   88s]         include yaml-0.11.5.0-G8RzVbeSOqB50sCYmn2TZ
[   88s]         include yesod-core-1.6.18.8-6NbLNd2S7d5Bf3HJp6rKid
[   88s]         include yesod-form-1.6.7-LC9WO26D5bjC0MNFROBEQe
[   88s]         include yesod-persistent-1.6.0.5-9cO5dk1kmHOAsa4tLTj0r1
[   88s]         Yesod=yesod-1.6.1.0-L1ZrEfHHjEL4ZPg5gUFJ40:Yesod,Yesod.Default.Config=yesod-1.6.1.0-L1ZrEfHHjEL4ZPg5gUFJ40:Yesod.Default.Config,Yesod.Default.Config2=yesod-1.6.1.0-L1ZrEfHHjEL4ZPg5gUFJ40:Yesod.Default.Config2,Yesod.Default.Handlers=yesod-1.6.1.0-L1ZrEfHHjEL4ZPg5gUFJ40:Yesod.Default.Handlers,Yesod.Default.Main=yesod-1.6.1.0-L1ZrEfHHjEL4ZPg5gUFJ40:Yesod.Default.Main,Yesod.Default.Util=yesod-1.6.1.0-L1ZrEfHHjEL4ZPg5gUFJ40:Yesod.Default.Util
[   88s] Ready component graph:
[   88s]     definite yesod-1.6.1.0-L1ZrEfHHjEL4ZPg5gUFJ40
[   88s]         depends aeson-1.5.6.0-DKRYsoqDhrXJNujTZrDHEf
[   88s]         depends base-4.15.0.0
[   88s]         depends bytestring-0.10.12.1
[   88s]         depends conduit-1.3.4.1-5jKrafZ5jYhKoJdTRMFD84
[   88s]         depends data-default-class-0.1.2.0-FEqadazbCfKLPgsOi8Mq6U
[   88s]         depends directory-1.3.6.1
[   88s]         depends fast-logger-3.0.3-FaKeK891u6TFBhkVRPQKLr
[   88s]         depends file-embed-0.0.13.0-4AN2M3x2UWSEsGAFWu7ZEd
[   88s]         depends monad-logger-0.3.36-FL2BtKw8EeaF4wZmehPP8a
[   88s]         depends shakespeare-2.0.25-4IGNqPAl1f13jGkgGXg0N5
[   88s]         depends streaming-commons-0.2.2.1-Hw3qCB3irjv4km2jUH6Jg0
[   88s]         depends template-haskell-2.17.0.0
[   88s]         depends text-1.2.4.1
[   88s]         depends unix-2.7.2.2
[   88s]         depends unordered-containers-0.2.13.0-LSqNoVnVBinKYmInNBh54X
[   88s]         depends wai-3.2.3-HxoYdmypAXX1MnCeAE2iqq
[   88s]         depends wai-extra-3.1.6-LudkXIwcaQS8Nqvm3IxzCV
[   88s]         depends wai-logger-2.3.6-I6z9W9PdRA25Ruoe0vixqu
[   88s]         depends warp-3.3.14-2I6laE4MgmWJYarfgyelkn
[   88s]         depends yaml-0.11.5.0-G8RzVbeSOqB50sCYmn2TZ
[   88s]         depends yesod-core-1.6.18.8-6NbLNd2S7d5Bf3HJp6rKid
[   88s]         depends yesod-form-1.6.7-LC9WO26D5bjC0MNFROBEQe
[   88s]         depends yesod-persistent-1.6.0.5-9cO5dk1kmHOAsa4tLTj0r1
[   88s] Using Cabal-3.4.0.0 compiled by ghc-9.0
[   88s] Using compiler: ghc-9.0.1
[   88s] Using install prefix: /usr
[   88s] Executables installed in: /usr/bin
[   88s] Libraries installed in: /usr/lib64/ghc-9.0.1/yesod-1.6.1.0
[   88s] Dynamic Libraries installed in: /usr/lib64/ghc-9.0.1/yesod-1.6.1.0
[   88s] Private executables installed in:
[   88s] /usr/libexec/x86_64-linux-ghc-9.0.1/yesod-1.6.1.0
[   88s] Data files installed in: /usr/share/yesod-1.6.1.0
[   88s] Documentation installed in: /usr/share/licenses/ghc-yesod
[   88s] Configuration files installed in: /usr/etc
[   88s] No alex found
[   88s] Using ar found on system at: /usr/bin/ar
[   88s] No c2hs found
[   88s] No cpphs found
[   88s] No doctest found
[   88s] Using gcc version 10 found on system at: /usr/bin/gcc
[   88s] Using ghc version 9.0.1 found on system at: /usr/bin/ghc
[   88s] Using ghc-pkg version 9.0.1 found on system at: /usr/bin/ghc-pkg
[   88s] No ghcjs found
[   88s] No ghcjs-pkg found
[   88s] No greencard found
[   88s] Using haddock version 2.24.0 found on system at: /usr/bin/haddock
[   88s] No happy found
[   88s] Using haskell-suite found on system at: haskell-suite-dummy-location
[   88s] Using haskell-suite-pkg found on system at: haskell-suite-pkg-dummy-location
[   88s] No hmake found
[   88s] Using hpc version 0.68 found on system at: /usr/bin/hpc
[   88s] Using hsc2hs version 0.68.7 found on system at: /usr/bin/hsc2hs
[   88s] No hscolour found
[   88s] No jhc found
[   88s] Using ld found on system at: /usr/bin/ld
[   88s] Using pkg-config version 1.7.3 found on system at: /usr/bin/pkg-config
[   88s] Using runghc version 9.0.1 found on system at: /usr/bin/runghc
[   88s] Using strip version 2.36 found on system at: /usr/bin/strip
[   88s] Using tar found on system at: /usr/bin/tar
[   88s] No uhc found
[   88s] + '[' -x Setup ']'
[   88s] + LANG=en_US.utf8
[   88s] + ./Setup build -v
[   88s] Component build order: library
[   88s] /usr/bin/ghc-pkg init dist/package.conf.inplace
[   88s] creating dist/build
[   88s] creating dist/build/autogen
[   88s] creating dist/build/autogen
[   88s] Preprocessing library for yesod-1.6.1.0..
[   88s] Building library for yesod-1.6.1.0..
[   88s] creating dist/build
[   88s] /usr/bin/ghc --make -fbuilding-cabal-package -O -static -dynamic-too -dynosuf dyn_o -dynhisuf dyn_hi -outputdir dist/build -odir dist/build -hidir dist/build -stubdir dist/build -i -idist/build -i. -idist/build/autogen -idist/build/global-autogen -Idist/build/autogen -Idist/build/global-autogen -Idist/build -optP-include -optPdist/build/autogen/cabal_macros.h -this-unit-id yesod-1.6.1.0-L1ZrEfHHjEL4ZPg5gUFJ40 -hide-all-packages -Wmissing-home-modules -no-user-package-db -package-db dist/package.conf.inplace -package-id aeson-1.5.6.0-DKRYsoqDhrXJNujTZrDHEf -package-id base-4.15.0.0 -package-id bytestring-0.10.12.1 -package-id conduit-1.3.4.1-5jKrafZ5jYhKoJdTRMFD84 -package-id data-default-class-0.1.2.0-FEqadazbCfKLPgsOi8Mq6U -package-id directory-1.3.6.1 -package-id fast-logger-3.0.3-FaKeK891u6TFBhkVRPQKLr -package-id file-embed-0.0.13.0-4AN2M3x2UWSEsGAFWu7ZEd -package-id monad-logger-0.3.36-FL2BtKw8EeaF4wZmehPP8a -package-id shakespeare-2.0.25-4IGNqPAl1f13jGkgGXg0N5 -package-id streaming-commons-0.2.2.1-Hw3qCB3irjv4km2jUH6Jg0 -package-id template-haskell-2.17.0.0 -package-id text-1.2.4.1 -package-id unix-2.7.2.2 -package-id unordered-containers-0.2.13.0-LSqNoVnVBinKYmInNBh54X -package-id wai-3.2.3-HxoYdmypAXX1MnCeAE2iqq -package-id wai-extra-3.1.6-LudkXIwcaQS8Nqvm3IxzCV -package-id wai-logger-2.3.6-I6z9W9PdRA25Ruoe0vixqu -package-id warp-3.3.14-2I6laE4MgmWJYarfgyelkn -package-id yaml-0.11.5.0-G8RzVbeSOqB50sCYmn2TZ -package-id yesod-core-1.6.18.8-6NbLNd2S7d5Bf3HJp6rKid -package-id yesod-form-1.6.7-LC9WO26D5bjC0MNFROBEQe -package-id yesod-persistent-1.6.0.5-9cO5dk1kmHOAsa4tLTj0r1 -XHaskell2010 Yesod Yesod.Default.Config Yesod.Default.Config2 Yesod.Default.Main Yesod.Default.Util Yesod.Default.Handlers -Wall -optc-O2 -optc-Wall '-optc-D_FORTIFY_SOURCE=2' -optc-fstack-protector-strong -optc-funwind-tables -optc-fasynchronous-unwind-tables -optc-fstack-clash-protection '-optc-Werror=return-type'
[   89s] [1 of 6] Compiling Yesod            ( Yesod.hs, dist/build/Yesod.o, dist/build/Yesod.dyn_o )
[   89s] [2 of 6] Compiling Yesod.Default.Config ( Yesod/Default/Config.hs, dist/build/Yesod/Default/Config.o, dist/build/Yesod/Default/Config.dyn_o )
[   90s] 
[   90s] Yesod/Default/Config.hs:191:14: warning: [-Wdeprecations]
[   90s]     In the use of ‘parseMonad’ (imported from Data.Yaml):
[   90s]     Deprecated: "With the MonadFail split, this function is going to be removed in the future. Please migrate to parseEither."
[   90s]     |
[   90s] 191 |     mport <- parseMonad (\x -> x .: "port") m
[   90s]     |              ^^^^^^^^^^
[   90s] 
[   90s] Yesod/Default/Config.hs:204:14: warning: [-Wdeprecations]
[   90s]     In the use of ‘parseMonad’ (imported from Data.Yaml):
[   90s]     Deprecated: "With the MonadFail split, this function is going to be removed in the future. Please migrate to parseEither."
[   90s]     |
[   90s] 204 |     extra <- parseMonad (parseExtra env) m
[   90s]     |              ^^^^^^^^^^
[   90s] 
[   90s] Yesod/Default/Config.hs:240:61: warning: [-Wdeprecations]
[   90s]     In the use of ‘parseMonad’ (imported from Data.Yaml):
[   90s]     Deprecated: "With the MonadFail split, this function is going to be removed in the future. Please migrate to parseEither."
[   90s]     |
[   90s] 240 |             | Just v <- M.lookup (T.pack $ show env) obj -> parseMonad f v
[   90s]     |                                                             ^^^^^^^^^^
[   91s] [3 of 6] Compiling Yesod.Default.Config2 ( Yesod/Default/Config2.hs, dist/build/Yesod/Default/Config2.o, dist/build/Yesod/Default/Config2.dyn_o )
[   92s] 
[   92s] Yesod/Default/Config2.hs:31:1: warning: [-Wunused-imports]
[   92s]     The import of ‘Data.Semigroup’ is redundant
[   92s]       except perhaps to import instances from ‘Data.Semigroup’
[   92s]     To import instances alone, use: import Data.Semigroup()
[   92s]    |
[   92s] 31 | import Data.Semigroup
[   92s]    | ^^^^^^^^^^^^^^^^^^^^^
[   93s] [4 of 6] Compiling Yesod.Default.Handlers ( Yesod/Default/Handlers.hs, dist/build/Yesod/Default/Handlers.o, dist/build/Yesod/Default/Handlers.dyn_o )
[   93s] [5 of 6] Compiling Yesod.Default.Main ( Yesod/Default/Main.hs, dist/build/Yesod/Default/Main.o, dist/build/Yesod/Default/Main.dyn_o )
[   93s] [6 of 6] Compiling Yesod.Default.Util ( Yesod/Default/Util.hs, dist/build/Yesod/Default/Util.o, dist/build/Yesod/Default/Util.dyn_o )
[   93s] 
[   93s] Yesod/Default/Util.hs:116:17: error:
[   93s]     • Couldn't match type ‘[Stmt] -> Exp’ with ‘Exp’
[   93s]       Expected: Q Exp
[   93s]         Actual: Q ([Stmt] -> Exp)
[   94s]     • In the expression: return $ DoE $ map NoBindS exps
[   94s]       In a case alternative: exps -> return $ DoE $ map NoBindS exps
[   94s]       In a stmt of a 'do' block:
[   94s]         case catMaybes mexps of
[   94s]           [] -> error $ concat ["Called ", func, ....]
[   94s]           exps -> return $ DoE $ map NoBindS exps
[   94s]     |
[   94s] 116 |         exps -> return $ DoE $ map NoBindS exps
[   94s]     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[   94s] 
[   94s] Yesod/Default/Util.hs:116:32: error:
[   94s]     • Couldn't match expected type: Maybe ModName
[   94s]                   with actual type: [Stmt]
[   94s]     • In the second argument of ‘($)’, namely ‘map NoBindS exps’
[   94s]       In the second argument of ‘($)’, namely ‘DoE $ map NoBindS exps’
[   94s]       In the expression: return $ DoE $ map NoBindS exps
[   94s]     |
[   94s] 116 |         exps -> return $ DoE $ map NoBindS exps
[   94s]     |                                ^^^^^^^^^^^^^^^^
```

**Comment by snoyberg** [comment]
Do you have a stack.yaml file this repros against?

**Comment by mimi1vx** [comment]
no. this is without the stack, and all module dependencies are in the first lines of the build. + `yesod-core` has applied patch from https://gitlab.haskell.org/ghc/head.hackage/-/blob/master/patches/yesod-core-1.6.18.8.patch

**Comment by snoyberg** [comment]
Closing as invalid

**Comment by mimi1vx** [comment]
oh , we can pretend as stack last compiler etc is ghc-8.10.4, in that case, is this report invalid. Or we can look outside the `stack` world and found there is stable GHC-9.0.1 and 9.0.2 coming, in this case, `yesod` isn't buildable

**Comment by snoyberg** [comment]
You opened a bug report using a patched version of this library without documenting it, built against patched versions of underlying libraries. That's an invalid bug report, and I'm not wasting my time trying to reproduce it.

**Comment by mimi1vx** [comment]
@snoyberg if you took 2 sec of your time .. mentioned patch is simple `conP` compatibility fix for https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0099-explicit-specificity.rst .. which is almost three years old proposal implemented in GHC-9.0+ and has nothing to do with this issue

**Comment by sheaf** [comment]
@mimi1vx Changing `DoE` to `DoE Nothing`, to account for the extra parameter added to `DoE` for the `QualifiedDo` extension, fixes the issue on my end and allows `yesod` to build. Sprinkle in some CPP for backwards compatibility.
